File: typescript/tests/callback-order.test.ts
/**
 * Tests for the callback execution order invariant: validate -> filter.
 *
 * The critical invariant is:
 *   1. Built-in validation runs first (validateAnswer)
 *   2. User-provided validate callback runs second
 *   3. Filter runs LAST, only on accepted values
 */

import { Readable, Writable } from "node:stream";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { resetAgent } from "../src/agent.js";
import { ValidationError } from "../src/errors.js";
import { CheckboxPrompt } from "../src/prompts/checkbox.js";
import { ExpandPrompt } from "../src/prompts/expand.js";
import { NumberPrompt } from "../src/prompts/number.js";
import { SelectPrompt } from "../src/prompts/select.js";
import { TextPrompt } from "../src/prompts/text.js";
import { resetSocketTransport } from "../src/socket.js";

const ACK = '{"kind":"handshake_ack"}';

function makeStdinFromLines(lines: string[]): Readable {
  const data = lines.map((l) => `${l}\n`).join("");
  return Readable.from(data);
}

function captureStdout(): { chunks: string[]; writable: Writable } {
  const chunks: string[] = [];
  const writable = new Writable({
    write(chunk: Buffer, _encoding: string, callback: () => void) {
      chunks.push(chunk.toString());
      callback();
    },
  });
  return { chunks, writable };
}

function setupAgentMode(answers: string[]) {
  resetAgent();
  resetSocketTransport();
  vi.stubEnv("INQUIRER_AI_MODE", "agent");
  vi.stubEnv("INQUIRER_AI_TRANSPORT", "stdio");

  const { writable } = captureStdout();
  const stdin = makeStdinFromLines([ACK, ...answers]);

  const origStdin = process.stdin;
  const origStdout = process.stdout;
  Object.defineProperty(process, "stdin", { value: stdin, configurable: true });
  Object.defineProperty(process, "stdout", { value: writable, configurable: true });

  return () => {
    Object.defineProperty(process, "stdin", { value: origStdin, configurable: true });
    Object.defineProperty(process, "stdout", { value: origStdout, configurable: true });
  };
}

describe("Callback order invariant: validate -> filter", () => {
  beforeEach(() => {
    resetAgent();
    resetSocketTransport();
  });

  // ── Test 1: Order invariant — validate is called before filter ──

  describe("Order invariant", () => {
    it("Text: validate runs before filter", async () => {
      const cleanup = setupAgentMode(['{"answer":"hello"}']);
      try {
        const log: string[] = [];
        const result = await new TextPrompt({
          message: "Q",
          validate: (_v: string) => { log.push("validate"); return true; },
          filter: (v: string) => { log.push("filter"); return v; },
        }).execute();
        expect(result).toBe("hello");
        expect(log).toEqual(["validate", "filter"]);
      } finally {
        cleanup();
      }
    });

    it("Number: validate runs before filter", async () => {
      const cleanup = setupAgentMode(['{"answer":42}']);
      try {
        const log: string[] = [];
        const result = await new NumberPrompt({
          message: "Q",
          validate: (_v: number) => { log.push("validate"); return true; },
          filter: (v: number) => { log.push("filter"); return v; },
        }).execute();
        expect(result).toBe(42);
        expect(log).toEqual(["validate", "filter"]);
      } finally {
        cleanup();
      }
    });

    it("Select: validate runs before filter", async () => {
      const cleanup = setupAgentMode(['{"answer":"a"}']);
      try {
        const log: string[] = [];
        const result = await new SelectPrompt({
          message: "Q",
          choices: [{ name: "A", value: "a" }],
          validate: (_v: unknown) => { log.push("validate"); return true; },
          filter: (v: unknown) => { log.push("filter"); return v; },
        }).execute();
        expect(result).toBe("a");
        expect(log).toEqual(["validate", "filter"]);
      } finally {
        cleanup();
      }
    });

    it("Expand: validate runs before filter", async () => {
      const cleanup = setupAgentMode(['{"answer":"y"}']);
      try {
        const log: string[] = [];
        const result = await new ExpandPrompt({
          message: "Q",
          choices: [{ key: "y", name: "Yes", value: "yes" }],
          validate: (_v: unknown) => { log.push("validate"); return true; },
          filter: (v: unknown) => { log.push("filter"); return v; },
        }).execute();
        expect(result).toBe("yes");
        expect(log).toEqual(["validate", "filter"]);
      } finally {
        cleanup();
      }
    });

    it("Checkbox: validate runs before filter", async () => {
      const cleanup = setupAgentMode(['{"answer":["a"]}']);
      try {
        const log: string[] = [];
        const result = await new CheckboxPrompt({
          message: "Q",
          choices: [{ name: "A", value: "a" }],
          validate: (_v: unknown[]) => { log.push("validate"); return true; },
          filter: (v: unknown[]) => { log.push("filter"); return v; },
        }).execute();
        expect(result).toEqual(["a"]);
        expect(log).toEqual(["validate", "filter"]);
      } finally {
        cleanup();
      }
    });
  });

  // ── Test 2: Filter NOT called when validate rejects ──

  describe("Filter skipped on rejection", () => {
    it("Text: filter not called when validate rejects", async () => {
      const cleanup = setupAgentMode([
        '{"answer":"bad"}',
        '{"answer":"bad"}',
        '{"answer":"bad"}',
        '{"answer":"bad"}',
      ]);
      try {
        const filterCalls: string[] = [];
        await expect(
          new TextPrompt({
            message: "Q",
            validate: () => "rejected",
            filter: (v: string) => { filterCalls.push(v); return v; },
          }).execute(),
        ).rejects.toThrow(ValidationError);
        expect(filterCalls).toEqual([]);
      } finally {
        cleanup();
      }
    });

    it("Number: filter not called when validate rejects", async () => {
      const cleanup = setupAgentMode([
        '{"answer":42}',
        '{"answer":42}',
        '{"answer":42}',
        '{"answer":42}',
      ]);
      try {
        const filterCalls: number[] = [];
        await expect(
          new NumberPrompt({
            message: "Q",
            validate: () => "rejected",
            filter: (v: number) => { filterCalls.push(v); return v; },
          }).execute(),
        ).rejects.toThrow(ValidationError);
        expect(filterCalls).toEqual([]);
      } finally {
        cleanup();
      }
    });

    it("Number: filter not called when built-in validation rejects (out of range)", async () => {
      const cleanup = setupAgentMode([
        '{"answer":100}',
        '{"answer":100}',
        '{"answer":100}',
        '{"answer":100}',
      ]);
      try {
        const filterCalls: number[] = [];
        const validateCalls: number[] = [];
        await expect(
          new NumberPrompt({
            message: "Q",
            min: 0,
            max: 10,
            validate: (v: number) => { validateCalls.push(v); return true; },
            filter: (v: number) => { filterCalls.push(v); return v; },
          }).execute(),
        ).rejects.toThrow(ValidationError);
        expect(filterCalls).toEqual([]);
        expect(validateCalls).toEqual([]);
      } finally {
        cleanup();
      }
    });
  });

  // ── Test 3: Filter receives raw value ──

  describe("Filter sees raw value", () => {
    it("Text: filter receives original value", async () => {
      const cleanup = setupAgentMode(['{"answer":"  HELLO  "}']);
      try {
        const received: string[] = [];
        const result = await new TextPrompt({
          message: "Q",
          validate: () => true,
          filter: (v: string) => { received.push(v); return v.trim(); },
        }).execute();
        expect(received).toEqual(["  HELLO  "]);
        expect(result).toBe("HELLO");
      } finally {
        cleanup();
      }
    });

    it("Number: filter receives validated number", async () => {
      const cleanup = setupAgentMode(['{"answer":42}']);
      try {
        const received: number[] = [];
        const result = await new NumberPrompt({
          message: "Q",
          validate: () => true,
          filter: (v: number) => { received.push(v); return v * 2; },
        }).execute();
        expect(received).toEqual([42]);
        expect(result).toBe(84);
      } finally {
        cleanup();
      }
    });
  });

  // ── Test 4: Multiple rejections, filter called only once ──

  describe("Multiple rejections", () => {
    it("Text: filter called only once after retries", async () => {
      const cleanup = setupAgentMode([
        '{"answer":"bad1"}',
        '{"answer":"bad2"}',
        '{"answer":"good"}',
      ]);
      try {
        const filterCalls: string[] = [];
        let attempt = 0;
        const result = await new TextPrompt({
          message: "Q",
          validate: (_v: string) => { attempt++; return attempt > 2 ? true : "rejected"; },
          filter: (v: string) => { filterCalls.push(v); return `${v}_ok`; },
        }).execute();
        expect(filterCalls).toEqual(["good"]);
        expect(result).toBe("good_ok");
      } finally {
        cleanup();
      }
    });

    it("Number: filter called only once after retries", async () => {
      const cleanup = setupAgentMode([
        '{"answer":1}',
        '{"answer":2}',
        '{"answer":3}',
      ]);
      try {
        const filterCalls: number[] = [];
        let attempt = 0;
        const result = await new NumberPrompt({
          message: "Q",
          validate: (_v: number) => { attempt++; return attempt > 2 ? true : "rejected"; },
          filter: (v: number) => { filterCalls.push(v); return v * 10; },
        }).execute();
        expect(filterCalls).toEqual([3]);
        expect(result).toBe(30);
      } finally {
        cleanup();
      }
    });

    it("Select: filter called only once after retries", async () => {
      const cleanup = setupAgentMode([
        '{"answer":"a"}',
        '{"answer":"a"}',
        '{"answer":"a"}',
      ]);
      try {
        const filterCalls: unknown[] = [];
        let attempt = 0;
        await new SelectPrompt({
          message: "Q",
          choices: [{ name: "A", value: "a" }],
          validate: (_v: unknown) => { attempt++; return attempt > 2 ? true : "rejected"; },
          filter: (v: unknown) => { filterCalls.push(v); return v; },
        }).execute();
        expect(filterCalls.length).toBe(1);
      } finally {
        cleanup();
      }
    });
  });

  // ── Test 5: Cross-type consistency ──

  describe("Cross-type consistency", () => {
    const cases = [
      {
        name: "Text",
        answer: '{"answer":"x"}',
        factory: (log: string[]) =>
          new TextPrompt({
            message: "Q",
            validate: () => { log.push("validate"); return true; },
            filter: (v: string) => { log.push("filter"); return v; },
          }),
      },
      {
        name: "Number",
        answer: '{"answer":5}',
        factory: (log: string[]) =>
          new NumberPrompt({
            message: "Q",
            validate: () => { log.push("validate"); return true; },
            filter: (v: number) => { log.push("filter"); return v; },
          }),
      },
      {
        name: "Select",
        answer: '{"answer":"x"}',
        factory: (log: string[]) =>
          new SelectPrompt({
            message: "Q",
            choices: [{ name: "X", value: "x" }],
            validate: () => { log.push("validate"); return true; },
            filter: (v: unknown) => { log.push("filter"); return v; },
          }),
      },
      {
        name: "Expand",
        answer: '{"answer":"y"}',
        factory: (log: string[]) =>
          new ExpandPrompt({
            message: "Q",
            choices: [{ key: "y", name: "Yes", value: "yes" }],
            validate: () => { log.push("validate"); return true; },
            filter: (v: unknown) => { log.push("filter"); return v; },
          }),
      },
      {
        name: "Checkbox",
        answer: '{"answer":["a"]}',
        factory: (log: string[]) =>
          new CheckboxPrompt({
            message: "Q",
            choices: [{ name: "A", value: "a" }],
            validate: () => { log.push("validate"); return true; },
            filter: (v: unknown[]) => { log.push("filter"); return v; },
          }),
      },
    ];

    for (const { name, answer, factory } of cases) {
      it(`${name}: validate -> filter order`, async () => {
        const cleanup = setupAgentMode([answer]);
        try {
          const log: string[] = [];
          await factory(log).execute();
          expect(log).toEqual(["validate", "filter"]);
        } finally {
          cleanup();
        }
      });
    }
  });
});

File: go/prompt/callback_order_test.go
package prompt

import (
	"fmt"
	"strings"
	"sync"
	"testing"
)

// ── Helpers ──

// callLog tracks the order of callback invocations.
type callLog struct {
	mu    sync.Mutex
	calls []string
}

func (c *callLog) record(name string) {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.calls = append(c.calls, name)
}

func (c *callLog) get() []string {
	c.mu.Lock()
	defer c.mu.Unlock()
	cp := make([]string, len(c.calls))
	copy(cp, c.calls)
	return cp
}

// ── Test 1: Order invariant — validate is called before filter ──

func TestTextCallbackOrder_ValidateBeforeFilter(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":"hello"}`+"\n")
	defer cleanup()

	log := &callLog{}
	_, err := Text(TextConfig{
		Message: "Q",
		Validate: func(s string) error {
			log.record("validate")
			return nil
		},
		Filter: func(s string) string {
			log.record("filter")
			return s
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	calls := log.get()
	if len(calls) != 2 {
		t.Fatalf("expected 2 calls, got %d: %v", len(calls), calls)
	}
	if calls[0] != "validate" || calls[1] != "filter" {
		t.Fatalf("expected [validate, filter], got %v", calls)
	}
}

func TestSelectCallbackOrder_ValidateBeforeFilter(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":"a"}`+"\n")
	defer cleanup()

	log := &callLog{}
	_, err := Select(SelectConfig{
		Message: "Q",
		Choices: []ChoiceItem{Choice{Name: "A", Value: "a"}},
		Validate: func(v any) error {
			log.record("validate")
			return nil
		},
		Filter: func(v any) any {
			log.record("filter")
			return v
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	calls := log.get()
	if len(calls) != 2 {
		t.Fatalf("expected 2 calls, got %d: %v", len(calls), calls)
	}
	if calls[0] != "validate" || calls[1] != "filter" {
		t.Fatalf("expected [validate, filter], got %v", calls)
	}
}

func TestCheckboxCallbackOrder_ValidateBeforeFilter(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":["a"]}`+"\n")
	defer cleanup()

	log := &callLog{}
	_, err := Checkbox(CheckboxConfig{
		Message: "Q",
		Choices: []ChoiceItem{Choice{Name: "A", Value: "a"}},
		Validate: func(v any) error {
			log.record("validate")
			return nil
		},
		Filter: func(v any) any {
			log.record("filter")
			return v
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	calls := log.get()
	if len(calls) != 2 {
		t.Fatalf("expected 2 calls, got %d: %v", len(calls), calls)
	}
	if calls[0] != "validate" || calls[1] != "filter" {
		t.Fatalf("expected [validate, filter], got %v", calls)
	}
}

func TestExpandCallbackOrder_ValidateBeforeFilter(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":"y"}`+"\n")
	defer cleanup()

	log := &callLog{}
	_, err := Expand(ExpandConfig{
		Message: "Q",
		Choices: []ExpandChoice{{Key: "y", Name: "Yes", Value: "yes"}},
		Validate: func(v any) error {
			log.record("validate")
			return nil
		},
		Filter: func(v any) any {
			log.record("filter")
			return v
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	calls := log.get()
	if len(calls) != 2 {
		t.Fatalf("expected 2 calls, got %d: %v", len(calls), calls)
	}
	if calls[0] != "validate" || calls[1] != "filter" {
		t.Fatalf("expected [validate, filter], got %v", calls)
	}
}

// ── Test 2: Filter NOT called when validate rejects ──

func TestTextFilterNotCalledOnRejection(t *testing.T) {
	// 3 rejections => exhausts retries
	input := strings.Repeat(`{"answer":"bad"}`+"\n", 3)
	r, w, cleanup := agentSetup(t, input)
	defer cleanup()

	filterCalled := false
	_, err := Text(TextConfig{
		Message: "Q",
		Validate: func(s string) error {
			return fmt.Errorf("rejected")
		},
		Filter: func(s string) string {
			filterCalled = true
			return s
		},
	})
	readOutput(r, w)

	if err == nil {
		t.Fatal("expected validation error")
	}
	if filterCalled {
		t.Fatal("filter should NOT be called when validate rejects")
	}
}

func TestSelectFilterNotCalledOnRejection(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":"a"}`+"\n")
	defer cleanup()

	filterCalled := false
	_, err := Select(SelectConfig{
		Message: "Q",
		Choices: []ChoiceItem{Choice{Name: "A", Value: "a"}},
		Validate: func(v any) error {
			return fmt.Errorf("rejected")
		},
		Filter: func(v any) any {
			filterCalled = true
			return v
		},
	})
	readOutput(r, w)

	if err == nil {
		t.Fatal("expected validation error")
	}
	if filterCalled {
		t.Fatal("filter should NOT be called when validate rejects")
	}
}

func TestCheckboxFilterNotCalledOnRejection(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":["a"]}`+"\n")
	defer cleanup()

	filterCalled := false
	_, err := Checkbox(CheckboxConfig{
		Message: "Q",
		Choices: []ChoiceItem{Choice{Name: "A", Value: "a"}},
		Validate: func(v any) error {
			return fmt.Errorf("rejected")
		},
		Filter: func(v any) any {
			filterCalled = true
			return v
		},
	})
	readOutput(r, w)

	if err == nil {
		t.Fatal("expected validation error")
	}
	if filterCalled {
		t.Fatal("filter should NOT be called when validate rejects")
	}
}

// ── Test 3: Filter receives raw (pre-validate) value ──

func TestTextFilterReceivesRawValue(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":"  RAW  "}`+"\n")
	defer cleanup()

	var filterInput string
	result, err := Text(TextConfig{
		Message: "Q",
		Validate: func(s string) error {
			// validate sees raw input
			if !strings.Contains(s, "RAW") {
				return fmt.Errorf("expected RAW")
			}
			return nil
		},
		Filter: func(s string) string {
			filterInput = s
			return strings.TrimSpace(s)
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	// filter receives the same raw value that validate saw
	if filterInput != "  RAW  " {
		t.Fatalf("filter should receive raw value '  RAW  ', got %q", filterInput)
	}
	// but result is trimmed by filter
	if result != "RAW" {
		t.Fatalf("expected trimmed 'RAW', got %q", result)
	}
}

// ── Test 4: Multiple rejections, filter called only on accepted value ──

func TestTextMultipleRejectionsFilterCalledOnce(t *testing.T) {
	input := `{"answer":"bad1"}` + "\n" +
		`{"answer":"bad2"}` + "\n" +
		`{"answer":"good"}` + "\n"
	r, w, cleanup := agentSetup(t, input)
	defer cleanup()

	filterCount := 0
	var filterValue string
	attempt := 0
	_, err := Text(TextConfig{
		Message: "Q",
		Validate: func(s string) error {
			attempt++
			if attempt <= 2 {
				return fmt.Errorf("not yet")
			}
			return nil
		},
		Filter: func(s string) string {
			filterCount++
			filterValue = s
			return s + "_filtered"
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if filterCount != 1 {
		t.Fatalf("filter should be called exactly once, was called %d times", filterCount)
	}
	if filterValue != "good" {
		t.Fatalf("filter should receive 'good', got %q", filterValue)
	}
}

func TestSelectMultipleRejectionsFilterCalledOnce(t *testing.T) {
	// Select only gets one shot per answer in agent mode (invalid choice error),
	// so we provide 3 answers where the first 2 are valid choices but validate rejects.
	input := `{"answer":"a"}` + "\n" +
		`{"answer":"a"}` + "\n" +
		`{"answer":"a"}` + "\n"
	r, w, cleanup := agentSetup(t, input)
	defer cleanup()

	filterCount := 0
	attempt := 0
	_, err := Select(SelectConfig{
		Message: "Q",
		Choices: []ChoiceItem{Choice{Name: "A", Value: "a"}},
		Validate: func(v any) error {
			attempt++
			if attempt <= 2 {
				return fmt.Errorf("not yet")
			}
			return nil
		},
		Filter: func(v any) any {
			filterCount++
			return v
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if filterCount != 1 {
		t.Fatalf("filter should be called exactly once, was called %d times", filterCount)
	}
}

// ── Test 5: Cross-type consistency — same order across Text, Select, Checkbox, Expand ──

func TestCrossTypeCallbackOrderConsistency(t *testing.T) {
	type testCase struct {
		name  string
		input string
		run   func(log *callLog) error
	}
	cases := []testCase{
		{
			name:  "Text",
			input: `{"answer":"hello"}` + "\n",
			run: func(log *callLog) error {
				_, err := Text(TextConfig{
					Message:  "Q",
					Validate: func(s string) error { log.record("validate"); return nil },
					Filter:   func(s string) string { log.record("filter"); return s },
				})
				return err
			},
		},
		{
			name:  "Select",
			input: `{"answer":"a"}` + "\n",
			run: func(log *callLog) error {
				_, err := Select(SelectConfig{
					Message:  "Q",
					Choices:  []ChoiceItem{Choice{Name: "A", Value: "a"}},
					Validate: func(v any) error { log.record("validate"); return nil },
					Filter:   func(v any) any { log.record("filter"); return v },
				})
				return err
			},
		},
		{
			name:  "Checkbox",
			input: `{"answer":["a"]}` + "\n",
			run: func(log *callLog) error {
				_, err := Checkbox(CheckboxConfig{
					Message:  "Q",
					Choices:  []ChoiceItem{Choice{Name: "A", Value: "a"}},
					Validate: func(v any) error { log.record("validate"); return nil },
					Filter:   func(v any) any { log.record("filter"); return v },
				})
				return err
			},
		},
		{
			name:  "Expand",
			input: `{"answer":"y"}` + "\n",
			run: func(log *callLog) error {
				_, err := Expand(ExpandConfig{
					Message:  "Q",
					Choices:  []ExpandChoice{{Key: "y", Name: "Yes", Value: "yes"}},
					Validate: func(v any) error { log.record("validate"); return nil },
					Filter:   func(v any) any { log.record("filter"); return v },
				})
				return err
			},
		},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			r, w, cleanup := agentSetup(t, tc.input)
			defer cleanup()

			log := &callLog{}
			if err := tc.run(log); err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			readOutput(r, w)

			calls := log.get()
			if len(calls) != 2 {
				t.Fatalf("expected 2 calls, got %d: %v", len(calls), calls)
			}
			if calls[0] != "validate" || calls[1] != "filter" {
				t.Fatalf("expected [validate, filter], got %v", calls)
			}
		})
	}
}

// ── Number-specific tests ──

func TestNumberCallbackOrder_ValidateBeforeFilter(t *testing.T) {
	r, w, cleanup := agentSetup(t, `{"answer":10}`+"\n")
	defer cleanup()

	log := &callLog{}
	_, err := Number(NumberConfig{
		Message: "Q",
		Validate: func(v float64) error {
			log.record("validate")
			return nil
		},
		Filter: func(v float64) float64 {
			log.record("filter")
			return v
		},
	})
	readOutput(r, w)

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	calls := log.get()
	if len(calls) != 2 {
		t.Fatalf("expected 2 calls, got %d: %v", len(calls), calls)
	}
	if calls[0] != "validate" || calls[1] != "filter" {
		t.Fatalf("expected [validate, filter], got %v", calls)
	}
}

func TestNumberFilterNotCalledOnBuiltInRejection(t *testing.T) {
	// Value exceeds max, so built-in validation rejects before filter/validate run
	min := 0.0
	max := 5.0
	r, w, cleanup := agentSetup(t, `{"answer":100}`+"\n")
	defer cleanup()

	filterCalled := false
	validateCalled := false
	_, err := Number(NumberConfig{
		Message: "Q",
		Min:     &min,
		Max:     &max,
		Validate: func(v float64) error {
			validateCalled = true
			return nil
		},
		Filter: func(v float64) float64 {
			filterCalled = true
			return v
		},
	})
	readOutput(r, w)

	if err == nil {
		t.Fatal("expected validation error for out-of-range value")
	}
	if filterCalled {
		t.Fatal("filter should NOT be called when built-in validation rejects")
	}
	if validateCalled {
		t.Fatal("user validate should NOT be called when built-in validation rejects")
	}
}
